StreamServerConnection does not notify the client about invalid messages

kkinnunen-apple opened the pull request:
#### a635e17acdc5b0cb49bc55649e2b159cf56f8f9e
<pre>
StreamServerConnection does not notify the client about invalid messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=297653">https://bugs.webkit.org/show_bug.cgi?id=297653</a>
<a href="https://rdar.apple.com/158761082">rdar://158761082</a>

Reviewed by Mike Wyrzykowski.

Add IPC::StreamServerConnection::Client interface that contains
IPC::StreamServerConnection::Client::didReceiveInvalidMessage(). This is
similar to IPC::Connection::Client.
The connection client interface is intended for the main holder of
the connection, the one that adds the connection to the runloop or
work queue as a &quot;event source&quot;. Currently the holders also receive
messages, so to be able to use CheckedPtr, the ::Client interface
is subclass of IPC::MessageReceiver/IPC::StreamMessageReceiver.

Add tests for both
 * IPC::StreamServerConnection::Client::didReceiveInvalidMessage
 * IPC::Connection::Client::didReceiveInvalidMessage

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::terminateWebProcess):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::terminateWebProcess):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::workQueueInitialize):
(WebKit::RemoteGraphicsContextGL::didReceiveInvalidMessage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLGBM.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLWC.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::workQueueInitialize):
(WebKit::RemoteRenderingBackend::didReceiveInvalidMessage):
(WebKit::RemoteRenderingBackend::terminateWebProcess): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::workQueueInitialize):
(WebKit::RemoteGPU::didReceiveInvalidMessage):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchMessage):
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.cpp:
(IPC::MessageReceiveQueueMap::get const):
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
(IPC::MessageReceiveQueueMap::isValidMessage): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::open):
(IPC::StreamServerConnection::invalidate):
(IPC::StreamServerConnection::didReceiveInvalidMessage):
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::processOutOfStreamMessage):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchDidReceiveInvalidMessage):
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::initialize):
(WebKit::IPCStreamTester::didReceiveInvalidMessage):
(WebKit::IPCStreamTester::checkInvalidMessages):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::LogStream):
(WebKit::LogStream::create):
(WebKit::LogStream::didReceiveInvalidMessage):
* Source/WebKit/Shared/LogStream.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createLogStream):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::serverIsA const):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::testType):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::openServer):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::openClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::server):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::client):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::serverClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::clientClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::deleteServer):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::deleteClient):
* Tools/TestWebKitAPI/Tests/IPC/EventTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.cpp:
(TestWebKitAPI::PrintTo):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::WaitForMessageMixin::~WaitForMessageMixin):
(TestWebKitAPI::WaitForMessageMixin::takeMessages):
(TestWebKitAPI::WaitForMessageMixin::takeInvalidMessages):
(TestWebKitAPI::WaitForMessageMixin::waitForMessage):
(TestWebKitAPI::WaitForMessageMixin::waitForDidClose):
(TestWebKitAPI::WaitForMessageMixin::gotDidClose const):
(TestWebKitAPI::WaitForMessageMixin::waitForInvalidMessage):
(TestWebKitAPI::WaitForMessageMixin::addMessage):
(TestWebKitAPI::WaitForMessageMixin::addInvalidMessage):
(TestWebKitAPI::WaitForMessageMixin::markDidClose):
(TestWebKitAPI::WaitForMessageMixin::WTF_GUARDED_BY_LOCK):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::StreamConnectionTestBase::setupBase): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::teardownBase): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::localReferenceBarrier): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::serverQueue): Deleted.
(TestWebKitAPI::TEST_F(StreamConnectionTest, OpenConnections)): Deleted.
(TestWebKitAPI::TEST_F(StreamConnectionTest, InvalidateUnopened)): Deleted.
(TestWebKitAPI::StreamMessageTest::StreamMessageTest): Deleted.
(TestWebKitAPI::StreamMessageTest::bufferSizeLog2 const): Deleted.
(TestWebKitAPI::StreamMessageTest::defaultDestinationID): Deleted.
(TestWebKitAPI::TEST_P): Deleted.

Canonical link: <a href="https://commits.webkit.org/299013@main">https://commits.webkit.org/299013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb5bae453c9fa6c54c79153016921b733e149fb5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27750 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123560 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69451 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9ca58b30-3405-402f-956e-b8d228758d2a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119338 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45717 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89141 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43798 "Build is being retried. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Exiting early after 60 failures. 21839 tests run. 60 failures; Uploaded test results; Exiting early after 60 failures. 21231 tests run. 60 failures; Compiled WebKit; layout-tests running") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a6df9255-58de-4c39-8ece-ae3203736010) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120412 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30115 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105322 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69650 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c280b950-0fbb-453f-ada8-5a871e455dd7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23436 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67234 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99520 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23618 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126680 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44357 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33347 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97807 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44715 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97601 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24850 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42971 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20909 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40708 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44230 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49889 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43687 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47031 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45382 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6631be50cfe89f4bdc311fe940bfac5c1f4b835b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49649

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1c5682d874ca550c7474384ac33643a07dae8f67. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49649

kkinnunen-apple committed:
StreamServerConnection does not notify the client about invalid messages
https://bugs.webkit.org/show_bug.cgi?id=297653
rdar://158761082

Reviewed by Mike Wyrzykowski.

Add IPC::StreamServerConnection::Client interface that contains
IPC::StreamServerConnection::Client::didReceiveInvalidMessage(). This is
similar to IPC::Connection::Client.
The connection client interface is intended for the main holder of
the connection, the one that adds the connection to the runloop or
work queue as a "event source". Currently the holders also receive
messages, so to be able to use CheckedPtr, the ::Client interface
is subclass of IPC::MessageReceiver/IPC::StreamMessageReceiver.

Add tests for both
 * IPC::StreamServerConnection::Client::didReceiveInvalidMessage
 * IPC::Connection::Client::didReceiveInvalidMessage

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::terminateWebProcess):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::terminateWebProcess):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::workQueueInitialize):
(WebKit::RemoteGraphicsContextGL::didReceiveInvalidMessage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLGBM.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLWC.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::workQueueInitialize):
(WebKit::RemoteRenderingBackend::didReceiveInvalidMessage):
(WebKit::RemoteRenderingBackend::terminateWebProcess): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::workQueueInitialize):
(WebKit::RemoteGPU::didReceiveInvalidMessage):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchMessage):
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.cpp:
(IPC::MessageReceiveQueueMap::get const):
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
(IPC::MessageReceiveQueueMap::isValidMessage): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::open):
(IPC::StreamServerConnection::invalidate):
(IPC::StreamServerConnection::didReceiveInvalidMessage):
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::processOutOfStreamMessage):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchDidReceiveInvalidMessage):
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::initialize):
(WebKit::IPCStreamTester::didReceiveInvalidMessage):
(WebKit::IPCStreamTester::checkInvalidMessages):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::LogStream):
(WebKit::LogStream::create):
(WebKit::LogStream::didReceiveInvalidMessage):
* Source/WebKit/Shared/LogStream.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createLogStream):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::serverIsA const):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::testType):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::openServer):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::openClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::server):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::client):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::serverClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::clientClient):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::deleteServer):
(TestWebKitAPI::ConnectionDidReceiveInvalidMessageTest::deleteClient):
* Tools/TestWebKitAPI/Tests/IPC/EventTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.cpp:
(TestWebKitAPI::PrintTo):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::WaitForMessageMixin::~WaitForMessageMixin):
(TestWebKitAPI::WaitForMessageMixin::takeMessages):
(TestWebKitAPI::WaitForMessageMixin::takeInvalidMessages):
(TestWebKitAPI::WaitForMessageMixin::waitForMessage):
(TestWebKitAPI::WaitForMessageMixin::waitForDidClose):
(TestWebKitAPI::WaitForMessageMixin::gotDidClose const):
(TestWebKitAPI::WaitForMessageMixin::waitForInvalidMessage):
(TestWebKitAPI::WaitForMessageMixin::addMessage):
(TestWebKitAPI::WaitForMessageMixin::addInvalidMessage):
(TestWebKitAPI::WaitForMessageMixin::markDidClose):
(TestWebKitAPI::WaitForMessageMixin::WTF_GUARDED_BY_LOCK):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::StreamConnectionTestBase::setupBase): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::teardownBase): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::localReferenceBarrier): Deleted.
(TestWebKitAPI::StreamConnectionTestBase::serverQueue): Deleted.
(TestWebKitAPI::TEST_F(StreamConnectionTest, OpenConnections)): Deleted.
(TestWebKitAPI::TEST_F(StreamConnectionTest, InvalidateUnopened)): Deleted.
(TestWebKitAPI::StreamMessageTest::StreamMessageTest): Deleted.
(TestWebKitAPI::StreamMessageTest::bufferSizeLog2 const): Deleted.
(TestWebKitAPI::StreamMessageTest::defaultDestinationID): Deleted.
(TestWebKitAPI::TEST_P): Deleted.

Canonical link: https://commits.webkit.org/299013@main
